tdrs-backend/tdpservice/parsers/case_consistency_validator.py

```diff
@@ -0,0 +1,127 @@
+"""Class definition for Category Four validator."""
+
+from .models import ParserErrorCategoryChoices
+from .util import get_rpt_month_year_list
+import logging
+
+logger = logging.getLogger(__name__)
+
+class CaseConsistencyValidator:
+    """Caches records of the same case to perform category four validation while actively parsing."""
+
+    def __init__(self, header, generate_error):
+        self.header = header
+        self.records = []
+        self.schemas = []
+        self.current_case = None
+        self.case_has_errors = False
+        self.section = header["type"]
+        self.case_is_section_one_or_two = self.section in {'A', 'C'}
+        self.program_type = header["program_type"]
+        self.has_validated = False
+        self.generate_error = generate_error
+        self.generated_errors = []
+        self.total_cases_cached = 0
+        self.total_cases_validated = 0
+
+    def clear_errors(self):
+        """Reset generated errors."""
+        self.generated_errors = []
+
+    def get_generated_errors(self):
+        """Return all errors generated for the current case."""
+        return self.generated_errors
+
+    def add_record(self, record, schema, case_has_errors):
+        """Add record to cache and validate if new case is detected."""
+        if self.case_is_section_one_or_two:
+            if record.CASE_NUMBER != self.current_case and self.current_case is not None:
+                self.validate()
+                self.records = [record]
+                self.schemas = [schema]
+                self.case_has_errors = case_has_errors
+            else:
+                self.case_has_errors = self.case_has_errors if self.case_has_errors else case_has_errors
+                self.records.append(record)
+                self.schemas.append(schema)
```

Comment:
a suggestion here is to append a tuple `(record, schema)` so that you don't have to manage two arrays or `zip` to loop both sets in the validator